tt_metal/hw/inc/dataflow_api.h

```diff
@@ -1537,40 +1561,143 @@ class Buffer {
+            pages_end = pages_start + num_pages_core;
+            uint32_t core_id_x = this->base_command_addr_[core_word_id + 1];
+            uint32_t core_id_y = this->base_command_addr_[core_word_id + 2];
+
+            //first get to correct core
+            if(!(page_id >= pages_start && page_id < pages_end)){
+                pages_start = pages_end;
+                continue;
+            }
+            core_id_start = core_word_id;
+            break;
+        }
+
+        uint32_t flattened_page_id = page_id;
+
+        uint32_t host_page_id = 0;
+        uint32_t host_offset = 0;
+        uint32_t core_page_id = (flattened_page_id - pages_start);
+        uint32_t core_offset = core_page_id * this->page_size_;
+
+        for(uint32_t core_word_id = core_id_start;  core_word_id < total_core_words; core_word_id+=NUM_ENTRIES_PER_SHARD){
+            uint32_t num_pages_core = this->base_command_addr_[core_word_id];
+            pages_end = pages_start + num_pages_core;
+            uint32_t core_id_x = this->base_command_addr_[core_word_id + 1];
+            uint32_t core_id_y = this->base_command_addr_[core_word_id + 2];
+
+
+            //now curr_page_id pointing to beginning of section we want in this core
+            uint32_t num_pages_write_core = min(pages_end - flattened_page_id, num_pages_left);
+
+            uint32_t size_in_bytes_written = num_pages_write_core * this->page_size_;
+
+            //Writing at beginning of core
+            uint64_t noc_address = get_noc_addr(core_id_x, core_id_y,
+                                            this->bank_base_address + core_offset);
+
+            if(!read){
+                noc_async_write(addr + host_offset, noc_address, size_in_bytes_written);
+            }
+            else{
+                noc_async_read(noc_address, addr + host_offset, size_in_bytes_written);
             }
+
+            num_pages_left-= num_pages_write_core;
+            host_offset += size_in_bytes_written;
+            if(num_pages_left == 0){
+                break;
+            }
+            core_offset = 0;
+            pages_start = pages_end;
+            flattened_page_id = pages_start;
         }
     }
 
-    void noc_async_read_buffer(uint32_t dst, const uint32_t id, const uint32_t num_pages, const uint32_t offset) {
-        if (this->type == BufferType::SYSTEM_MEMORY) {
-            noc_async_read(this->get_noc_addr(id, offset), dst, this->page_size_ * num_pages);
-        } else {
-            for (uint32_t i = 0; i < num_pages; i++) {
-                uint64_t address = this->get_noc_addr(id + i, offset);
-                noc_async_read(address, dst, this->page_size_);
-                dst += this->page_size_;
+    void noc_async_write_buffer(uint32_t src, const uint32_t id, const uint32_t num_pages, const uint32_t offset=0) {
+        if (this->sharded) {
+            noc_async_sharded_read_write_helper(src, false, num_pages, id);
```

Comment:
we should definitely templatize noc_async_sharded_read_write_helper instead of passing in read/write as a runtime argument.